cupy/_statistics/order.py

```diff
@@ -245,10 +245,17 @@ def _quantile_unchecked(a, q, axis=None, out=None, interpolation='linear',
             ptrdiff_t idx_below = floor(idx);
             U weight_above = idx - idx_below;
 
-            ptrdiff_t offset_i = _ind.get()[0] * offset;
             ptrdiff_t max_idx = size - 1;
-            ret = a[offset_i + idx_below] * (1.0 - weight_above)
-              + a[min(offset_i + idx_below + 1, max_idx)] * weight_above;
+            ptrdiff_t offset_bottom = _ind.get()[0] * offset + idx_below;
+            ptrdiff_t offset_top = min(offset_bottom + 1, max_idx);
+
+            U diff = a[offset_top] - a[offset_bottom];
+
+            if (weight_above < 0.5) {
+                ret = a[offset_bottom] + diff * weight_above;
+            } else {
+                ret = a[offset_top] - diff * (1 - weight_above);
+            }
```

Comment:
It looks like this branching would cause thread divergence. Could you benchmark against a few cases and report the timing before and after this patch? You can use `cupyx.time.repeat()` for a quick one-liner timing.

Reply:
Good thing to check! Here are the results for master:
```
For array of length 5
f                   :    CPU:  280.502 us   +/- 3.425 (min:  274.539 / max:  320.130) us     GPU-0:  420.902 us   +/-18.950 (min:  410.592 / max:  524.288) us

For array of length 50
f                   :    CPU:  279.530 us   +/- 3.893 (min:  273.767 / max:  489.012) us     GPU-0:  415.905 us   +/- 2.777 (min:  358.080 / max:  539.712) us

For array of length 500
f                   :    CPU:  279.705 us   +/- 3.187 (min:  273.260 / max:  404.803) us     GPU-0:  415.872 us   +/- 2.557 (min:  411.648 / max:  536.448) us

For array of length 5000
f                   :    CPU:  279.608 us   +/- 3.561 (min:  274.327 / max:  459.135) us     GPU-0:  415.901 us   +/- 2.545 (min:  411.648 / max:  504.128) us

For array of length 50000
f                   :    CPU:  279.674 us   +/- 3.171 (min:  273.374 / max:  371.667) us     GPU-0:  415.424 us   +/- 2.578 (min:  410.208 / max:  528.384) us

For array of length 500000000
f                   :    CPU:  279.140 us   +/- 3.060 (min:  274.150 / max:  336.556) us     GPU-0:  415.554 us   +/- 2.390 (min:  411.648 / max:  468.448) us
```
and here are results for this PR:
```
For array of length 5
f                   :    CPU:  281.037 us   +/- 4.982 (min:  275.461 / max:  411.347) us     GPU-0:  418.918 us   +/-18.364 (min:  410.624 / max:  528.576) us

For array of length 50
f                   :    CPU:  278.693 us   +/- 4.214 (min:  273.581 / max:  479.576) us     GPU-0:  414.290 us   +/- 2.819 (min:  409.792 / max:  530.112) us

For array of length 500
f                   :    CPU:  280.481 us   +/-21.542 (min:  273.537 / max:  893.934) us     GPU-0:  415.439 us   +/-15.847 (min:  410.080 / max:  909.312) us

For array of length 5000
f                   :    CPU:  280.989 us   +/- 3.962 (min:  275.230 / max:  428.586) us     GPU-0:  416.624 us   +/- 2.904 (min:  357.152 / max:  553.696) us

For array of length 50000
f                   :    CPU:  281.072 us   +/- 3.541 (min:  275.484 / max:  429.908) us     GPU-0:  416.318 us   +/- 2.795 (min:  411.936 / max:  563.200) us

For array of length 500000000
f                   :    CPU:  281.569 us   +/- 3.258 (min:  275.810 / max:  379.555) us     GPU-0:  416.871 us   +/- 2.502 (min:  412.320 / max:  513.376) us
```
Performance seems pretty comparable, which is what I would expect given how lightweight the work in each branch is, but let me know if you want me to try any other benchmarks or more rigorous tests.

Reply:
Thanks! Sorry I dropped the ball. Is this done with a random array in [0, 1] with q=0.5?

Reply:
Terribly sorry, when I went back to look at this, I realized there was a typo in my benchmarking script, so the size was not getting reset properly in between runs. The results are still quite comparable between master and the PR, though, even with the correction. I uploaded the script I used to [this gist](https://gist.github.com/wphicks/774701358b3aa4be33277aa4cc7cd7d2). It computes the 0, 50th, and 100th percentile for randomly generated arrays in [0, 1].

Here are the updated results:

master
```
For array of length 5
f                   :    CPU:  241.075 us   +/- 3.559 (min:  236.662 / max:  336.358) us     GPU-0:  246.124 us   +/- 3.660 (min:  240.576 / max:  342.400) us

For array of length 50
f                   :    CPU:  240.851 us   +/- 3.512 (min:  236.594 / max:  391.911) us     GPU-0:  245.895 us   +/- 3.601 (min:  241.600 / max:  397.312) us

For array of length 500
f                   :    CPU:  241.476 us   +/- 2.963 (min:  237.328 / max:  326.025) us     GPU-0:  246.539 us   +/- 3.059 (min:  241.888 / max:  332.320) us

For array of length 5000
f                   :    CPU:  279.828 us   +/- 3.995 (min:  274.140 / max:  402.477) us     GPU-0:  419.932 us   +/-18.860 (min:  356.352 / max:  518.144) us

For array of length 50000
f                   :    CPU:  279.626 us   +/- 4.327 (min:  273.721 / max:  405.583) us     GPU-0:  416.089 us   +/- 2.952 (min:  411.360 / max:  488.512) us

For array of length 5000000
f                   :    CPU:  284.788 us   +/-13.464 (min:  276.029 / max:  969.155) us     GPU-0: 2994.113 us   +/- 9.859 (min: 2926.592 / max: 3420.000) us
```

This PR
```
For array of length 5 
f                   :    CPU:  238.881 us   +/- 3.213 (min:  234.596 / max:  361.869) us     GPU-0:  243.750 us   +/- 3.312 (min:  238.624 / max:  368.832) us
                                                                                                       
For array of length 50                                                                                 
f                   :    CPU:  239.363 us   +/- 2.571 (min:  235.609 / max:  284.523) us     GPU-0:  244.241 us   +/- 2.673 (min:  239.648 / max:  290.816) us
                                                                                                       
For array of length 500                                                                                
f                   :    CPU:  240.826 us   +/- 3.010 (min:  236.789 / max:  316.428) us     GPU-0:  245.730 us   +/- 3.502 (min:  241.664 / max:  405.536) us
                                                                                                       
For array of length 5000                                                                               
f                   :    CPU:  276.527 us   +/- 3.105 (min:  271.654 / max:  317.652) us     GPU-0:  424.910 us   +/-15.261 (min:  417.792 / max:  517.696) us
                                                                                                       
For array of length 50000                                                                              
f                   :    CPU:  277.168 us   +/- 3.408 (min:  271.612 / max:  388.030) us     GPU-0:  422.038 us   +/- 2.751 (min:  417.824 / max:  530.432) us
                                                                                                       
For array of length 5000000                                                                            
f                   :    CPU:  280.512 us   +/- 3.707 (min:  273.256 / max:  356.554) us     GPU-0: 2993.909 us   +/- 5.918 (min: 2926.240 / max: 3227.808) us
```

Reply:
Nice, thanks!